implement rop.cache to store and return cached ROP search results

devalgupta404 opened the pull request:
**Your checklist for this pull request**
  - [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md) to this repository.
  - [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#code-style).
  - [x] I've documented every `RZ_API` function and struct this PR changes.
  - [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
  - [x] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).
  - [x] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

  **Detailed description**

  The `rop.cache` config option existed but was never implemented. The `cache` field in `RzRopSearchContext` was set from config but never used in the search logic.

  This PR implements the caching functionality by:
  - Using the existing `ht_rop` hashtable in `RzAnalysis` to store search results
  - Keying results by DJB2 hash of the grep argument
  - Returning cached results immediately on subsequent identical searches
  - Properly freeing the hashtable on analysis cleanup


  **Test plan**

  1. Unit tests pass:
  meson test -C build rop rop_constraint
  1/2 rizin:unit / rop            OK
  2/2 rizin:unit / rop_constraint OK

  2. Timing proof (same as issue #5544):
  [0x00000000]> e rop.cache=true
  [0x00000000]> time /R/ jalr
    0x000001fa               2ad6  sw a0, 44(sp)
    0x000001fc               6df2  bnez a2, 0x1de
    0x000001fe               0293  jalr t1
    ...
  0.056576
  [0x00000000]> time /R/ jalr
    0x000001fa               2ad6  sw a0, 44(sp)
    0x000001fc               6df2  bnez a2, 0x1de
    0x000001fe               0293  jalr t1
    ...
  0.000016

  | Search | Time | Notes |
  |--------|------|-------|
  | 1st | 0.056576s | Actual search |
  | 2nd | 0.000016s | Cached (~3500x faster) |

  **Closing issues**

  closes #5544

devalgupta404 committed:
implement rop.cache to store and return cached ROP search results

Rot127 commented on test/unit/test_rop.c:
```diff
@@ -141,8 +141,27 @@ bool test_rz_direct_solver() {
+		core, "mov", false, RZ_ROP_GADGET_PRINT, RZ_ROP_DETAIL_SEARCH_NON, NULL);
+	RzCmdStatus status = rz_core_rop_search(core, context);
+	mu_assert_eq(status, RZ_CMD_STATUS_OK, "rop search should succeed");
+	mu_assert_notnull(core->analysis->ht_rop, "ht_rop should be initialized after cached search");
```
Please also check if the correct result is in the cache.

Then do another search for the same gadget.

devalgupta404 replied:
@Rot127 should i manually insert result into cache and then verify the second search returns the cached result correctly?

Rot127 replied:
No, after the search the result is expected to be in the cache.

devalgupta404 committed:
update test to check if cached result is correct or not

devalgupta404 committed:
Test verifies that ROP search results are cached and reused on subsequent searches

devalgupta404 commented:
@Rot127 did you check the updated test?

Rot127 commented:
Please fix the build first

devalgupta404 committed:
fix build issue

devalgupta404 committed:
Added platform-specific binary selection in the test

devalgupta404 commented:
@Rot127 i fixed the build issue please take a look.

Rot127 commented:
Tests still fail though. Please check

devalgupta404 committed:
Fix rop.cache to not cache empty greparg searches

devalgupta404 committed:
Fix clang-format issues in test_rop.c

devalgupta404 committed:
Update test expectations for improved analysis output

- ARM: Update jump table case label from '3...4' to '4' (more precise)
- x86_64: Update switch/case ranges to match improved analysis output
- cmd_print: Update histogram EXPECT to include ANSI codes that rizin outputs

devalgupta404 committed:
Revert "Update test expectations for improved analysis output"

This reverts commit acca26610a494311626da0d000b029b254b77a33.

devalgupta404 committed:
fix: memory leaks in ROP search and unit tests

devalgupta404 committed:
fix(test): resolve hello-linux-x86_64 path dynamically

devalgupta404 commented:
@Rot127 i fixed the build issue and test locally whatever was failing previously
## Local Test Verification

### Fix: [test_rop](cci:1://file:///home/deval/python/rizin/test/unit/test_rop.c:143:0-203:1) binary path resolution

Updated [test/unit/test_rop.c](cci:7://file:///home/deval/python/rizin/test/unit/test_rop.c:0:0-0:0) to correctly resolve the test binary path.

### Results

**test_rop:**

$ build/test/unit/test_rop
test_rop_cache OK
test_rz_direct_solver OK


**meson test:**

Ok:                 146 
Fail:               0   


**rz-test:**

24068 OK    1048 BR    11 XX    25 FX

(11 XX are pre-existing failures unrelated to this PR)

So can you please rerun the workflows

devalgupta404 committed:
move test_rop_cache to integration test suite

devalgupta404 commented:
@Rot127                                                                                                                                                                                       
                                                                                                                                                                                                
  While running CI tests locally, I found memory leaks but noticed that CI already                                                                                                              
  tolerates them (`.github/workflows/ci.yml` lines 106, 125, 143 - `detect_leaks=0`                                                                                                             
  and `allow_failure: true`).                                                                                                                                                                   
                                                                                                                                                                                                
  I fixed some leaks anyway - **should I keep or revert these changes?**                                                                                                                        
                                                                                                                                                                                                
  Also, I moved some tests from unit to integration after learning that unit tests                                                                                                              
  shouldn't use binaries. Is this correct?
